### PR TITLE
Wordpress installation and configutation with Ansible and Vagrant

### PR DESCRIPTION
**Pull request by danpaldev**

### CODE REVIEW

File: wordpress.yaml [hunk 1]
```diff
@@ -1,7 +1,136 @@
...
+      become: true
+      apt:
+        name: 
+          - mysql-server
+          - python3-pymysql  
+        state: present
+    
+    # Set MySQL root Password on Ansible/Ubuntu - https://linuxhint.com/set_mysql_root_password_ansible/ 
+    - name: Relax the MySQL root user
+      shell: sudo mysql -u root -e 'UPDATE mysql.user SET plugin="mysql_native_password"
```
Comment (korney4eg) [inline wordpress.yaml]: here you use `sudo` why?

File: wordpress.yaml [hunk 2]
```diff
@@ -1,7 +1,136 @@
...
+      tags:
+        - wp
+        - install  
+
+    - name: Checking if Wordpress is installed
+      stat: path=/var/www/html/wordpress
+      register: wp_installed
+
+    - name: Install Wordpress
+      shell: "sudo wp core download --path=/var/www/html/wordpress --allow-root"
```
Comment (korney4eg) [inline wordpress.yaml]: why `sudo`?

File: wordpress.yaml [hunk 3]
```diff
@@ -1,7 +1,136 @@
...
+      when: not wp_installed.stat.exists
+
+    - name: Verifying if wp-config file exists
+      stat: path=/var/www/html/wordpress/wp-config.php
+      register: wp_configured
+
+    - name: Configure Wordpress DB Connection
+      shell: "sudo wp config create --dbname=wp_db --dbuser=wp_user --dbpass=wp_password --path=/var/www/html/wordpress --allow-root"
```
Comment (korney4eg) [inline wordpress.yaml]: why `sudo`?

File: wordpress.yaml [hunk 4]
```diff
@@ -1,7 +1,136 @@
...
+      when: not wp_configured.stat.exists
+
+    - name: Configure Wordpress Admin User
+      shell: 'sudo wp core install --url="192.168.50.2" --title="Blogging Platform" --admin_user=admin --admin_email=danpal@example.com --admin_password="!2three456." --path=/var/www/html/wordpress --allow-root'
```
Comment (korney4eg) [inline wordpress.yaml]: why `sudo`?

File: wordpress.yaml [hunk 5]
```diff
@@ -1,7 +1,136 @@
...
+  
+    - name: Activating twentynineteen wp theme
+      shell: 'sudo wp theme install twentynineteen --activate --path=/var/www/html/wordpress --allow-root'
```
Comment (korney4eg) [inline wordpress.yaml]: why `sudo`?

File: wordpress.yaml [hunk 6]
```diff
@@ -1,7 +1,136 @@
...
+          - php-xml
+          - php-gd
+          - php-curl
+        state: present
+     
+    - name: Install nginx
+      become: true
+      apt:
+        name: nginx 
+        state: present
```
Comment (korney4eg) [inline wordpress.yaml:30]: after full setup could you restart your VM and check if wordpress is running?

File: wordpress.yaml [hunk 7]
```diff
@@ -1,7 +1,136 @@
...
+      ufw:
+        rule: allow
+        name: 'Nginx Full'
+    
+    - name: Copy NGINX default settings file
+      become: true
+      copy:
+        src: /vagrant/default
+        dest: /etc/nginx/sites-enabled
+        force: yes
```
Comment (korney4eg) [inline wordpress.yaml:43]: doesn't look like `nginx` will automatically get your new configuration

File: wordpress.yaml [hunk 8]
```diff
@@ -1,7 +1,136 @@
...
+
+    - name: Install MySQL Server
+      become: true
+      apt:
+        name: 
+          - mysql-server
+          - python3-pymysql
```
Comment (korney4eg) [inline wordpress.yaml:50]: why do you need `python3-pymysql`?

File: wordpress.yaml [hunk 9]
```diff
@@ -1,7 +1,136 @@
...
+          - python3-pymysql  
+        state: present
+    
+    # Set MySQL root Password on Ansible/Ubuntu - https://linuxhint.com/set_mysql_root_password_ansible/ 
+    - name: Relax the MySQL root user
+      shell: sudo mysql -u root -e 'UPDATE mysql.user SET plugin="mysql_native_password"
+       WHERE user="root" AND host="localhost"'
+
+    - name: Flush Privileges
+      shell: sudo mysql -u root -e 'FLUSH PRIVILEGES'
```
Comment (korney4eg) [inline wordpress.yaml]: here you use `sudo` why?

Comment (korney4eg) [reply wordpress.yaml]: can you put Flush privileges to previous command to make it single command?

File: wordpress.yaml [hunk 10]
```diff
@@ -1,7 +1,136 @@
...
+        login_user: 'root'
+        login_password: ''
+        name: wp_user
+        password: wp_password
+        priv: "*.*:ALL"
+        state: present
+     
+    - name: Create DB for wordpress
+      mysql_db:
+        login_host: 'localhost'
```
Comment (korney4eg) [inline wordpress.yaml]: what does mean `login_host`?

File: wordpress.yaml (hunk 2 above)
Comment (korney4eg) [reply wordpress.yaml]: also it's always recommend to run commands as non-root user

File: wordpress.yaml (hunk 3 above)
Comment (korney4eg) [reply wordpress.yaml]: why not copying existing config.php file or not generating from template?

File: wordpress.yaml [hunk 11]
```diff
@@ -1,7 +1,136 @@
...
+
+    - name: Giving Write permissions to the Wordpress Folder
+      become: true
+      shell: "chown -R www-data: /var/www/html/wordpress" 
+
+    - name: Refresh nginx
+      become: true
+      systemd: 
+        name: nginx
+        state: reloaded
```
Comment (korney4eg) [inline wordpress.yaml:133]: you reload nginx every time you run provision. Do you really need it?

File: default [hunk 12]
```diff
@@ -0,0 +1,22 @@
+server {
+	listen 80 default_server;
+	listen [::]:80 default_server;
+
+	root /var/www/html/wordpress;
+
+	index index.php index.html index.htm;
+
+	#server_name localhost;
+	server_name 192.168.50.2;
```
Comment (korney4eg) [inline default:10]: what is the reason to use `server_name` ?

**Review by korney4eg — COMMENTED** [review]
It looks like working, but some things would be nice to correct.

File: wordpress.yaml (hunk 1 above)
Comment (danpaldev) [reply wordpress.yaml]: I was having troubles using the 'root' user without adding `sudo`. For some reason mysql required it. Although I could try to remove it and use `become: true`.

File: wordpress.yaml (hunk 2 above)
Comment (danpaldev) [reply wordpress.yaml]: As I'm downloading the wordpress core inside a protected subdirectory, root permissions were required. wp-cli advices against using root and that's why I use the `--allow-root` flag.

I could try to don't use wp-cli as root but I would have to add more tasks.

File: wordpress.yaml (hunk 3 above)
Comment (danpaldev) [reply wordpress.yaml]: I could have reused the config.php file from task 03 but wanted to try the `wp config create` command. I could use the file if required tho.

File: wordpress.yaml (hunk 8 above)
Comment (danpaldev) [reply wordpress.yaml:50]: In order to use the ansible sql commands `mysql_db` and `mysql_user` I had to install `python3-pymysql` (Suggested also in the docs)

File: default (hunk 12 above)
Comment (danpaldev) [reply default:10]: `server_name` points to the `url` specified when setting up wordpress `wp core install --url="192.168.50.2"`. 

At first I didn't specify this property but routes and theme of wp didn't work.

File: wordpress.yaml (hunk 1 above)
Comment (korney4eg) [reply wordpress.yaml]: that was my point

File: wordpress.yaml (hunk 3 above)
Comment (korney4eg) [reply wordpress.yaml]: having same file would allow you to remove some steps to check if config exists

File: wordpress.yaml (hunk 8 above)
Comment (korney4eg) [reply wordpress.yaml:50]: ok, make sence

File: wordpress.yaml (hunk 11 above)
Comment (danpaldev) [reply wordpress.yaml:133]: I tried to remove it, and also changing its position, but this is the only place where it works. I'm still researching if there's a better way to do this.

File: wordpress.yaml (hunk 10 above)
Comment (danpaldev) [reply wordpress.yaml]: I was trying to follow _verbatim_ the Ansible docs and the login host was included in all examples. It made sense at the time. However, I actually tried the tasks without this and it seems that this property defaults to 'localhost' if not specified, so it's redundant indeed.

File: wordpress.yaml (hunk 6 above)
Comment (danpaldev) [reply wordpress.yaml:30]: I tried a `vagrant halt` and `vagrant up` to check this. WP runs correctly and with no data loss after a restart.

**Pull request closed without merging**